tensorflow/python/ops/math_ops.py

```diff
@@ -2041,48 +2041,38 @@ def accumulate_n(inputs, shape=None, tensor_dtype=None, name=None):
     ValueError: If `inputs` don't all have same shape and dtype or the shape
     cannot be inferred.
   """
+  _INPUTS_ERR_MSG = ValueError("inputs must be a list of at least one Tensor"
+                               "with the same dtype and shape")
   if context.in_eager_mode():
     # TODO(apassos) remove this once the lifetime of eager variables gets
     # addressed.
     raise ValueError("accumulate_n not supported in eager mode")
   if not inputs or not isinstance(inputs, (list, tuple)):
-    raise ValueError("inputs must be a list of at least one Tensor with the "
-                     "same dtype and shape")
+    raise _INPUTS_ERR_MSG
   inputs = ops.convert_n_to_tensor_or_indexed_slices(inputs)
   if not all(isinstance(x, ops.Tensor) for x in inputs):
-    raise ValueError("inputs must be a list of at least one Tensor with the "
-                     "same dtype and shape")
+    raise _INPUTS_ERR_MSG
   if not all(x.dtype == inputs[0].dtype for x in inputs):
-    raise ValueError("inputs must be a list of at least one Tensor with the "
-                     "same dtype and shape")
+    raise _INPUTS_ERR_MSG
   if shape is not None:
     shape = tensor_shape.as_shape(shape)
   else:
     shape = tensor_shape.unknown_shape()
   for input_tensor in inputs:
     if isinstance(input_tensor, ops.Tensor):
       shape = shape.merge_with(input_tensor.get_shape())
-  if tensor_dtype is None:
-    tensor_dtype = inputs[0].dtype
-  if tensor_dtype != inputs[0].dtype:
+
+  # tensor_dtype is for safety only; operator's output type computed in C++
+  if tensor_dtype is not None and tensor_dtype != inputs[0].dtype:
     raise TypeError("tensor_dtype is {}, but input is of type {}"
                     .format(tensor_dtype, inputs[0].dtype))
-  if len(inputs) == 1:
+
+  if len(inputs) == 1 and name is None:
     return inputs[0]
-  with ops.name_scope(name, "AccumulateN", inputs) as name:
```

Comment:
Can we separate the change which adds the C++ infrastructure from the change which switches the default implementation of accumulate_n?

For now, let's add the new op in contrib somewhere.

Tensorflow tries to maintain forward compatibility (new graphs should work on C++ runtimes which are two weeks old) and adding the new op and switching to use it at the same time breaks it.

Reply:
Sure, no problem. I can restructure the code so the new implementation is a different op with a different name and the old op remains unchanged.